src/main/java/com/artipie/scripting/ScriptRunner.java

```diff
@@ -0,0 +1,244 @@
+                            Logger.error(
+                                ScriptRunner.class,
+                                "Execution error in script %s %[exception]s",
+                                key.toString(),
+                                exc
+                            );
+                            result = Optional.empty();
+                        }
+                        return result;
+                    }
+                );
+        } else {
+            Logger.warn(ScriptRunner.class, "Cannot find script %s", key.toString());
+        }
+    }
+
+    /**
+     * Stops the job and logs error.
+     * @param context Job context
+     */
+    private void stopJob(final JobExecutionContext context) {
+        final JobKey key = context.getJobDetail().getKey();
+        try {
+            Logger.error(this, String.format("Force stopping job %s...", key));
+            new StdSchedulerFactory().getScheduler().deleteJob(key);
+            Logger.error(this, String.format("Job %s stopped.", key));
+        } catch (final SchedulerException error) {
+            Logger.error(this, String.format("Error while stopping job %s", key));
+            throw new ArtipieException(error);
+        }
+    }
+
+    /**
+     * Returns cached script. Scripts objects are handled by LoadingCache.
+     * @param key Key in the `storage` for the script.
+     * @param storage Targer storage object.
+     * @return Returns cached script or empty Optional in case of error.
+     */
+    private static Optional<Script> cachedScript(final Key key, final BlockingStorage storage) {
+        final String ext = ScriptRunner.extension(key.string());
+        final Script.ScriptType type = Arrays.stream(Script.ScriptType.values())
+            .filter(val -> val.ext().equals(ext)).findFirst().orElse(Script.ScriptType.NONE);
+        Optional<Script> result = Optional.empty();
+        if (!type.equals(Script.ScriptType.NONE)) {
+            final ScriptRunner.FilesContent content = new ScriptRunner.FilesContent(
+                key, storage
+            );
+            result = Optional.ofNullable(ScriptRunner.scripts.getUnchecked(content));
+        }
+        return result;
+    }
+
+    /**
+     * Obtain extension of filename.
+     * @param filename Name of file.
+     * @return Extension.
+     */
+    private static String extension(final String filename) {
+        final int pos = filename.lastIndexOf('.');
+        final String res;
+        if (pos >= 0) {
+            res = filename.substring(pos + 1);
+        } else {
+            res = "";
+        }
+        return res;
+    }
+
+    /**
+     * Extra class for caching precompiled scripts.
+     * @since 0.1
+     */
+    private static final class FilesContent {
+        /**
+         * Key.
+         */
+        private final Key key;
+
+        /**
+         * Storage.
+         */
+        private final BlockingStorage storage;
```

Comment:
Why do you include `storage` in the key!? 
This doesn't seem right, don't do this.

Reply:
@dgarus , currently `key` strongly depends on correct `storage`, since storage objects could be the set of storage wrappers, etc. Here `ScriptRunner` wraps `configStorage` with `BlockingStorage` from the `settings` passed in `JobDataMap`. 
For these reasons I need either holding `storage` instance in addition to the key, or read whole script and use script source code in the `String` form as the key for precompiled scripts cache.

Reply:
> @dgarus , currently `key` strongly depends on correct `storage`, since storage objects could be the set of storage wrappers, etc. Here `ScriptRunner` wraps `configStorage` with `BlockingStorage` from the `settings` passed in `JobDataMap`. For these reasons I need either holding `storage` instance in addition to the key, or read whole script and use script source code in the `String` form as the key for precompiled scripts cache.

Storage cannot be a part of key. This is wrong!

Reply:
@dgarus , storage will be removed from `equals()/hashCode()`, but kept inside key for data access.